mempool/clist_mempool.go

```diff
@@ -649,29 +610,139 @@ func (mem *CListMempool) Update(
 	return nil
 }
 
+// recheckTxs sends all transactions in the mempool to the app for re-validation. When the function
+// returns, all recheck responses from the app have been processed.
```

Comment:
This is not about this PR, but blocking here is terrible. Why we need to conclude the recheck before move on? And, if this is indeed the case, why we use `CheckTxAsync`, expected to be... asynchronous.

Reply:
It's not blocking, that's why there is a timeout, so that rechecking can finish even when not all the CheckTx responses have arrived.

Remember that recheck happens after the mempool has been updated. And once rechecking finishes then the `updateMtx` lock is released and adding new transactions via `CListMempool.CheckTx` is possible again.

Reply:
Sorry, ignore above when I say it's non-blocking. Rechecking needs to block waiting for all reCheckTx responses because it needs to finish before new transactions are allowed to be checked again. This is to not break the sequential order when rechecking transactions.

Reply:
Bear in mind that, even if we wait here until all TXs have been re-Checked, it still makes sense that `CheckTx` is async: in non-local ABCI clients we will pipeline this:
* Request serialization, request transmission (over TCP or gRPC), Request deserialization at the app, etc
If we make `CheckTx` synchronous, all that will have to happen for one TX until we start with the next. This may make no difference with local client, but is likely to have a high performance impact over a socket or gRPC

Reply:
@cason In order not to have to wait here, we need to tackle one of the items in our laundry list: `Re-CheckTx`

But, as you said, that's not part of this PR